File: source/doctor.ts
/**
 * @copyright Copyright (c) 2026 GUIHO Technologies as represented by Cristóvão GUIHO. All Rights Reserved.
 */

import { stat } from 'node:fs/promises'
import { basename, join, relative, resolve } from 'node:path'
import type { XDocsConfig, XDocsDoctorIssue, XDocsDoctorOptions, XDocsDoctorResult } from './types.js'
import { scanProject } from './discovery.js'
import { scanMetadata } from './meta.js'
import { validateTree } from './tree.js'

export {
  doctorProject,
}

/** Run strict xdocs health checks for descriptors, tree links, documents, and file references. */
const doctorProject = async (config: XDocsConfig, options: XDocsDoctorOptions = {}): Promise<XDocsDoctorResult> => {
  const targetPath = resolve(config.cwd, options.targetPath ?? config.cwd)
  const issues: XDocsDoctorIssue[] = []
  const scan = await scanProject(config)
  const metadata = await scanMetadata(config, { targetPath: options.targetPath, includeDocuments: options.includeDocuments ?? true })
  const treeValidation = validateTree(scan.xdocsFiles)
  const targetPrefix = relative(config.cwd, targetPath)

  for (const file of scan.xdocsFiles.filter((entry) => isInScope(entry.relativePath, targetPrefix))) {
    for (const error of file.errors) {
      issues.push({ severity: 'error', code: 'descriptor-invalid', path: file.relativePath, message: error })
    }
  }

  for (const descriptor of metadata.descriptors) {
    for (const error of descriptor.errors) {
      issues.push({ severity: 'error', code: 'metadata-invalid', path: descriptor.relativePath, message: error })
    }

    for (const document of descriptor.documents) {
      for (const error of document.errors) {
        issues.push({
          severity: options.warningsAsErrors ? 'error' : 'warning',
          code: 'document-metadata',
          path: document.relativePath,
          message: error,
        })
      }
    }
  }

  for (const error of treeValidation.errors) {
    issues.push({ severity: 'error', code: 'tree-invalid', path: null, message: error })
  }

  for (const warning of treeValidation.warnings) {
    issues.push({ severity: options.warningsAsErrors ? 'error' : 'warning', code: 'tree-warning', path: null, message: warning })
  }

  for (const descriptor of metadata.descriptors) {
    if (!descriptor.metadata) continue

    for (const fileName of Object.keys(descriptor.metadata.files)) {
      if (!isSiblingFileName(fileName)) {
        issues.push({ severity: 'error', code: 'file-entry-invalid', path: descriptor.relativePath, message: `Invalid file entry: "${fileName}" must be a sibling filename, not a path.` })
        continue
      }

      const filePath = join(descriptor.directory, fileName)
      const fileStat = await stat(filePath).catch(() => null)
      if (!fileStat?.isFile()) {
        issues.push({ severity: 'error', code: 'file-missing', path: descriptor.relativePath, message: `Missing documented file: "${fileName}" is listed in metadata but does not exist beside the descriptor.` })
      }
    }
  }

  const dedupedIssues = dedupeIssues(issues)
  const errors = dedupedIssues.filter((issue) => issue.severity === 'error').length
  const warnings = dedupedIssues.filter((issue) => issue.severity === 'warning').length

  return {
    root: config.cwd,
    targetPath: relative(config.cwd, targetPath) || '.',
    valid: errors === 0,
    summary: { errors, warnings },
    issues: dedupedIssues,
  }
}

const isInScope = (relativePath: string, targetPrefix: string): boolean => {
  if (!targetPrefix) return true
  return relativePath === targetPrefix || relativePath.startsWith(`${targetPrefix}\\`) || relativePath.startsWith(`${targetPrefix}/`)
}

const isSiblingFileName = (name: string): boolean =>
  Boolean(name) && basename(name) === name && !name.includes('/') && !name.includes('\\')

const dedupeIssues = (issues: XDocsDoctorIssue[]): XDocsDoctorIssue[] => {
  const seen = new Set<string>()
  const result: XDocsDoctorIssue[] = []

  for (const issue of issues) {
    const key = `${issue.severity}\0${issue.code}\0${issue.path ?? ''}\0${issue.message}`
    if (seen.has(key)) continue
    seen.add(key)
    result.push(issue)
  }

  return result.sort((left, right) => {
    if (left.severity !== right.severity) return left.severity === 'error' ? -1 : 1
    return `${left.path ?? ''}${left.message}`.localeCompare(`${right.path ?? ''}${right.message}`)
  })
}

File: internal/xdocs/doctor.go
package xdocs

import (
	"fmt"
	"os"
	"path/filepath"
	"sort"
	"strings"

	"github.com/CGuiho/xdocs/internal/config"
)

func Doctor(cfg config.Config, options DoctorOptions) (DoctorResult, error) {
	target := cfg.CWD
	if options.TargetPath != "" {
		target = options.TargetPath
		if !filepath.IsAbs(target) {
			target = filepath.Join(cfg.CWD, target)
		}
	}
	target, _ = filepath.Abs(target)
	scan, err := ScanProject(cfg)
	if err != nil {
		return DoctorResult{}, err
	}
	meta, err := ScanMetadata(cfg, MetaOptions{
		TargetPath:       options.TargetPath,
		IncludeDocuments: options.IncludeDocuments,
	})
	if err != nil {
		return DoctorResult{}, err
	}
	issues := []DoctorIssue{}
	for _, file := range scan.XDocsFiles {
		if !inScope(file.Path, target) {
			continue
		}
		path := file.RelativePath
		for _, message := range file.Errors {
			issues = append(issues, DoctorIssue{Severity: "error", Code: "descriptor-invalid", Path: &path, Message: message})
		}
	}
	for _, descriptor := range meta.Descriptors {
		path := descriptor.RelativePath
		for _, message := range descriptor.Errors {
			issues = append(issues, DoctorIssue{Severity: "error", Code: "metadata-invalid", Path: &path, Message: message})
		}
		for _, document := range descriptor.Documents {
			documentPath := document.RelativePath
			severity := "warning"
			if options.WarningsAsErrors {
				severity = "error"
			}
			for _, message := range document.Errors {
				issues = append(issues, DoctorIssue{Severity: severity, Code: "document-metadata", Path: &documentPath, Message: message})
			}
		}
		if descriptor.Metadata == nil {
			continue
		}
		for name := range descriptor.Metadata.Files {
			if filepath.Base(name) != name || strings.ContainsAny(name, `/\`) {
				issues = append(issues, DoctorIssue{Severity: "error", Code: "file-entry-invalid", Path: &path, Message: fmt.Sprintf(`Invalid file entry: "%s" must be a sibling filename, not a path.`, name)})
				continue
			}
			if info, err := os.Stat(filepath.Join(descriptor.Directory, name)); err != nil || !info.Mode().IsRegular() {
				issues = append(issues, DoctorIssue{Severity: "error", Code: "file-missing", Path: &path, Message: fmt.Sprintf(`Missing documented file: "%s" is listed in metadata but does not exist beside the descriptor.`, name)})
			}
		}
	}
	validation := ValidateTree(scan.XDocsFiles)
	for _, message := range validation.Errors {
		issues = append(issues, DoctorIssue{Severity: "error", Code: "tree-invalid", Message: message})
	}
	for _, message := range validation.Warnings {
		severity := "warning"
		if options.WarningsAsErrors {
			severity = "error"
		}
		issues = append(issues, DoctorIssue{Severity: severity, Code: "tree-warning", Message: message})
	}
	issues = dedupeIssues(issues)
	summary := DoctorSummary{}
	for _, issue := range issues {
		if issue.Severity == "error" {
			summary.Errors++
		} else {
			summary.Warnings++
		}
	}
	return DoctorResult{
		Root: cfg.CWD, TargetPath: displayTarget(cfg.CWD, target),
		Valid: summary.Errors == 0, Summary: summary, Issues: issues,
	}, nil
}

func dedupeIssues(issues []DoctorIssue) []DoctorIssue {
	seen := map[string]bool{}
	result := []DoctorIssue{}
	for _, issue := range issues {
		path := ""
		if issue.Path != nil {
			path = *issue.Path
		}
		key := issue.Severity + "\x00" + issue.Code + "\x00" + path + "\x00" + issue.Message
		if !seen[key] {
			seen[key] = true
			result = append(result, issue)
		}
	}
	sort.Slice(result, func(i, j int) bool {
		if result[i].Severity != result[j].Severity {
			return result[i].Severity == "error"
		}
		left, right := result[i].Message, result[j].Message
		if result[i].Path != nil {
			left = *result[i].Path + left
		}
		if result[j].Path != nil {
			right = *result[j].Path + right
		}
		return left < right
	})
	return result
}
